### ISSUE TITLE
`string` should be `string?` for optional properties if `"generateOptionalPropertiesAsNullable": true`

### ISSUE THREAD

**Issue by SymboLinker** [RicoSuter/NSwag#4280]
NSwagStudio v13.16.1.0

Runtime: Net60
Outputs: CSharpClient
Settings:
"generateOptionalPropertiesAsNullable": true,
"generateNullableReferenceTypes": true,

The Newtonsoft.Json.Required.DisallowNull enum value means:
`The property is not required but it cannot be a null value.`
https://www.newtonsoft.com/json/help/html/t_newtonsoft_json_required.htm
so then it is "optional".

Expected Generated Outputs (part):
```
        [Newtonsoft.Json.JsonProperty("repaymentType", Required = Newtonsoft.Json.Required.DisallowNull, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
        public string? RepaymentType { get; set; } = default!;
```
Actual Generated Outputs (part):
```
        [Newtonsoft.Json.JsonProperty("repaymentType", Required = Newtonsoft.Json.Required.DisallowNull, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
        public string RepaymentType { get; set; } = default!;
```

For nullable reference types it looks OK:
```
        [Newtonsoft.Json.JsonProperty("installmentAmount", Required = Newtonsoft.Json.Required.DisallowNull, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
        public MoneyAmount? InstallmentAmount { get; set; } = default!;
```

**Comment by markm77** [comment]
Yeah, I noticed this for strings and DateTimeOffset values, very frustrating.

**Comment by TWhidden** [comment]
+1.

**Comment by RicoSuter** [comment]
Might be fixed with the next version of NJS/NSwag.

**Comment by omarmciver** [comment]
Did this issue go away? I'm finding if I use `generateNullableReferenceTypes=true` with `generateOptionalPropertiesAsNullable=true` then my C# client ends up setting a lot of required properties as Nullable.

When I stand back and think about this, if I'm using Nullable correctly in my source (I am generating Swagger from an existing .NET Controller app) then using these together makes no sense. By expecting my source to define nullable (and therefore optional) properties explicitly, why would I ever use `generateOptionalPropertiesAsNullable`?

This has been a head scratcher. Can anyone straighten out my thinking?

**Comment by markm77** [comment]
>By expecting my source to define nullable (and therefore optional) properties explicitly, why would I ever use generateOptionalPropertiesAsNullable?

Hi @omarmciver ,

I'm not sure I completely understand your post (and don't have bandwidth to debug your code problem) but let me have a go at answering your question.

If you have an API described by an OpenAPI document, JSON properties can be described as optional (i.e. not required) and/or nullable.

These are two different things but either or both can be used for a JSON property whose value doesn't need to be specified. (APIs I use tend to use optional rather than nullable for this purpose (why waste bandwidth sending "null"?).)

You then need to think about the mapping between C# and JSON. If you receive JSON null, you probably want this to map to C# null. If you receive (absence of optional value) you probably also want this to map to C# null (and not the C# default value which is probably a valid value). How else would you represent (absence of optional value) in C#?

So the combination of `GenerateNullableReferenceTypes` and `GenerateOptionalPropertiesAsNullable` is extremely useful for correct mapping between JSON and C#.

BR,
Mark

**Comment by omarmciver** [comment]
Hey @markm77 . Thank you for the explanation. I shouldn't have asked this on a Friday! I'll try to digest next week. I know my config is now working how I expect it to, I just haven't fully flushed out in my mind as to why :)

I appreciate your time and understand the bandwidth challenges. If I can contribute back in a way that helps, I'll be sure to PR against the documentation.
